File: src/main/kotlin/cz/vondr/idea/debugging/s050_exception_breakpoint/ExceptionBreakpoint.java
package cz.vondr.idea.debugging.s050_exception_breakpoint;

// Idea is, that its often can be good idea to enable Exception Breakpoint
// Then we cannot miss any exception
// BUT we have to properly setup filter

public class ExceptionBreakpoint {

    public static void main(String[] args) {
        startThreadThrowingNumberFormatExceptions();

        while (true) {
            try {
                Thread.sleep(5000);
                throw new SampleApplicationException("Interesting application exception");
            } catch (SampleApplicationException e) {
                // Exception handled
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt(); // Handle interrupt exception properly
            }
        }
    }

    public static void startThreadThrowingNumberFormatExceptions() {
        Thread thread = new Thread(() -> {
            while (true) {
                try {
                    Thread.sleep(100);
                    throw new NumberFormatException("Not interesting internal exception used to handle valid state");
                } catch (NumberFormatException e) {
                    // Exception handled
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt(); // Handle interrupt exception properly
                }
            }
        });
        thread.start();
    }

    public static class SampleApplicationException extends Exception {
        public SampleApplicationException(String message) {
            super(message);
        }
    }
}

File: src/main/kotlin/cz/vondr/idea/debugging/s050_exception_breakpoint/ExceptionBreakpoint.kt
package cz.vondr.idea.debugging.s050_exception_breakpoint

fun main() {
    startThreadThrowingNumberFormatExceptions()

    while (true) {
        Thread.sleep(5000)
        try {
            throw SampleApplicationException("Interesting application excepiton")
        } catch (e: SampleApplicationException) {
            /* Exception handled */
        }
    }

}

fun startThreadThrowingNumberFormatExceptions() {
    Thread {
        while (true) {
            Thread.sleep(100)
            try {
                throw NumberFormatException("Not interesting internal exception used to handle valid state")
            } catch (e: NumberFormatException) {
                /* Exception handled */
            }
        }
    }.start()
}

class SampleApplicationException(message: String?) : Exception(message)